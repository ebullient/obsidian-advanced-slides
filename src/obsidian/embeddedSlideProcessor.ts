import type { SlidesExtendedPlugin } from "../slidesExtended-Plugin";
import type { EmbeddedSlideParameters } from "../@types";
import { load } from "js-yaml";

export class EmbeddedSlideProcessor {
    private plugin: SlidesExtendedPlugin;
    constructor(plugin: SlidesExtendedPlugin) {
        this.plugin = plugin;
    }

    handler = async (source: string, el: HTMLElement) => {
        try {
            const parameters = this.readParameters(source);
            const page = parameters.page ? `${parameters.page}` : "0";

            const url = new URL(
                `http://localhost:${this.plugin.settings.port}/embed/${parameters.slide}#/${page}`,
            );
            url.searchParams.append("embed", "true");

            const viewContent = el.createDiv();

            viewContent.empty();
            viewContent.addClass("reveal-preview-view");

            viewContent.createEl("iframe", {
                attr: {
                    src: url.toString(),
                    sandbox: "allow-scripts allow-same-origin allow-popups",
                },
            });
        } catch (e) {
            el.createEl("h2", { text: `Parameters invalid: ${e.message}` });
        }
    };

    readParameters(src: string): EmbeddedSlideParameters {
        const params = load(src) as EmbeddedSlideParameters;
        const slide = this.plugin.obsidianUtils.findFile(
            params.slide.toString(),
        );
        if (slide?.endsWith(".md")) {
            params.slide = slide;
        } else if (slide) {
            params.slide = `${slide}.md`;
        }
        return params;
    }
}
