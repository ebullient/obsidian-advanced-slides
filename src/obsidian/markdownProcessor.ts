import type { Options } from "../@types";
import { YamlStore } from "../yaml/yamlStore";
import type { ObsidianUtils } from "./obsidianUtils";
import { AutoClosingProcessor } from "./processors/autoClosingProcessor";
import { BlockProcessor } from "./processors/blockProcessor";
import { CalloutProcessor } from "./processors/calloutProcessor";
import { ChartProcessor } from "./processors/chartProcessor";
import { CommentProcessor } from "./processors/commentProcessor";
import { DebugViewProcessor } from "./processors/debugViewProcessor";
import { DefaultBackgroundProcessor } from "./processors/defaultBackgroundProcessor";
import { DropProcessor } from "./processors/dropProcessor";
import { EmojiProcessor } from "./processors/emojiProcessor";
import { ExcalidrawProcessor } from "./processors/excalidrawProcessor";
import { FootnoteProcessor } from "./processors/footNoteProcessor";
import { FormatProcessor } from "./processors/formatProcessor";
import { FragmentProcessor } from "./processors/fragmentProcessor";
import { GridProcessor } from "./processors/gridProcessor";
import { IconsProcessor } from "./processors/iconsProcessor";
import { InternalLinkProcessor } from "./processors/internalLinkProcessor";
import { LatexProcessor } from "./processors/latexProcessor";
import { MediaProcessor } from "./processors/mediaProcessor";
import { MermaidProcessor } from "./processors/mermaidProcessor";
import { MultipleFileProcessor } from "./processors/multipleFileProcessor";
import { ReferenceProcessor } from "./processors/referenceProcessor";
import { SkipSlideProcessor } from "./processors/skipSlideProcessor";
import { TemplateProcessor } from "./processors/templateProcessor";

export class MarkdownProcessor {
    private multipleFileProcessor: MultipleFileProcessor;
    private blockProcessor: BlockProcessor;
    private mediaProcessor: MediaProcessor;
    private internalLinkProcessor: InternalLinkProcessor;
    private footnoteProcessor: FootnoteProcessor;
    private latexProcessor: LatexProcessor;
    private formatProcessor: FormatProcessor;
    private excalidrawProcessor: ExcalidrawProcessor;
    private mermaidProcessor: MermaidProcessor;
    private fragmentProcessor: FragmentProcessor;
    private gridProcessor: GridProcessor;
    private commentProcessor: CommentProcessor;
    private dropProcessor: DropProcessor;
    private autoClosingProcessor: AutoClosingProcessor;
    private emojiProcessor: EmojiProcessor;
    private iconsProcessor: IconsProcessor;
    private debugViewProcessor: DebugViewProcessor;
    private calloutProcessor: CalloutProcessor;
    private templateProcessor: TemplateProcessor;
    private chartProcessor: ChartProcessor;
    private defaultBackgroundProcessor: DefaultBackgroundProcessor;
    private referenceProcessor: ReferenceProcessor;
    private skipSlideProcessor: SkipSlideProcessor;

    constructor(utils: ObsidianUtils) {
        this.multipleFileProcessor = new MultipleFileProcessor(utils);
        this.blockProcessor = new BlockProcessor();
        this.mediaProcessor = new MediaProcessor(utils);
        this.internalLinkProcessor = new InternalLinkProcessor(utils);
        this.footnoteProcessor = new FootnoteProcessor();
        this.latexProcessor = new LatexProcessor();
        this.formatProcessor = new FormatProcessor();
        this.excalidrawProcessor = new ExcalidrawProcessor(utils);
        this.mermaidProcessor = new MermaidProcessor();
        this.fragmentProcessor = new FragmentProcessor();
        this.gridProcessor = new GridProcessor();
        this.commentProcessor = new CommentProcessor();
        this.dropProcessor = new DropProcessor();
        this.autoClosingProcessor = new AutoClosingProcessor();
        this.emojiProcessor = new EmojiProcessor();
        this.iconsProcessor = new IconsProcessor();
        this.debugViewProcessor = new DebugViewProcessor();
        this.calloutProcessor = new CalloutProcessor();
        this.templateProcessor = new TemplateProcessor(utils);
        this.chartProcessor = new ChartProcessor();
        this.defaultBackgroundProcessor = new DefaultBackgroundProcessor();
        this.referenceProcessor = new ReferenceProcessor();
        this.skipSlideProcessor = new SkipSlideProcessor();
    }

    process(markdown: string, options: Options) {
        YamlStore.getInstance().options = options;

        let before = this.trimEnding(markdown, options);
        let after: string;

        let circuitCounter = 0;
        while (before !== after) {
            circuitCounter++;
            if (after) {
                before = after;
            }
            const afterMultipleFileProcessor =
                this.multipleFileProcessor.process(before);
            after = this.templateProcessor.process(
                afterMultipleFileProcessor,
                options,
            );

            //Remove default templates after first pass
            options.defaultTemplate = null;

            if (circuitCounter > 9) {
                console.warn(
                    "WARNING: Circuit in template hierarchy detected!",
                );
                break;
            }
        }

        const afterSkipSlideProcessor = this.skipSlideProcessor.process(
            after,
            options,
        );
        const afterReferenceProcessor = this.referenceProcessor.process(
            afterSkipSlideProcessor,
        );
        const afterDebugViewProcessor = this.debugViewProcessor.process(
            afterReferenceProcessor,
            options,
        );
        const afterAutoClosingProcessor = this.autoClosingProcessor.process(
            afterDebugViewProcessor,
        );
        const defaultBackgroundProcessor =
            this.defaultBackgroundProcessor.process(
                afterAutoClosingProcessor,
                options,
            );
        const afterCalloutProcessor = this.calloutProcessor.process(
            defaultBackgroundProcessor,
        );
        const afterEmojiProcessor = this.emojiProcessor.process(
            afterCalloutProcessor,
        );
        const afterIconsProcessor =
            this.iconsProcessor.process(afterEmojiProcessor);
        const afterMermaidProcessor =
            this.mermaidProcessor.process(afterIconsProcessor);
        const afterBlockProcessor = this.blockProcessor.process(
            afterMermaidProcessor,
        );
        const afterFootNoteProcessor = this.footnoteProcessor.process(
            afterBlockProcessor,
            options,
        );
        const afterExcalidrawProcessor = this.excalidrawProcessor.process(
            afterFootNoteProcessor,
        );
        const afterMediaProcessor = this.mediaProcessor.process(
            afterExcalidrawProcessor,
        );
        const afterInternalLinkProcessor = this.internalLinkProcessor.process(
            afterMediaProcessor,
            options,
        );
        const afterLatexProcessor = this.latexProcessor.process(
            afterInternalLinkProcessor,
        );
        const afterFormatProcessor =
            this.formatProcessor.process(afterLatexProcessor);
        const afterFragmentProcessor = this.fragmentProcessor.process(
            afterFormatProcessor,
            options,
        );
        const afterDropProcessor = this.dropProcessor.process(
            afterFragmentProcessor,
            options,
        );
        const afterGridProcessor = this.gridProcessor.process(
            afterDropProcessor,
            options,
        );
        const afterCommentProcessor =
            this.commentProcessor.process(afterGridProcessor);
        const afterChartProcessor = this.chartProcessor.process(
            afterCommentProcessor,
            options,
        );

        if (options.log) {
            this.log("markdown", "", markdown);
            this.log("merge & template", markdown, after);
            this.log("afterSkipSlideProcessor", after, afterSkipSlideProcessor);
            this.log(
                "afterReferenceProcessor",
                afterSkipSlideProcessor,
                afterReferenceProcessor,
            );
            this.log(
                "afterDebugViewProcessor",
                afterReferenceProcessor,
                afterDebugViewProcessor,
            );
            this.log(
                "afterAutoClosingProcessor",
                afterDebugViewProcessor,
                afterAutoClosingProcessor,
            );
            this.log(
                "defaultBackgroundProcessor",
                afterAutoClosingProcessor,
                defaultBackgroundProcessor,
            );
            this.log(
                "afterCalloutProcessor",
                defaultBackgroundProcessor,
                afterCalloutProcessor,
            );
            this.log(
                "afterEmojiProcessor",
                afterCalloutProcessor,
                afterEmojiProcessor,
            );
            this.log(
                "afterIconsProcessor",
                afterEmojiProcessor,
                afterIconsProcessor,
            );
            this.log(
                "afterMermaidProcessor",
                afterIconsProcessor,
                afterMermaidProcessor,
            );
            this.log(
                "afterBlockProcessor",
                afterMermaidProcessor,
                afterBlockProcessor,
            );
            this.log(
                "afterFootNoteProcessor",
                afterBlockProcessor,
                afterFootNoteProcessor,
            );
            this.log(
                "afterExcalidrawProcessor",
                afterFootNoteProcessor,
                afterExcalidrawProcessor,
            );
            this.log(
                "afterMediaProcessor",
                afterExcalidrawProcessor,
                afterMediaProcessor,
            );
            this.log(
                "afterInternalLinkProcessor",
                afterMediaProcessor,
                afterInternalLinkProcessor,
            );
            this.log(
                "afterLatexProcessor",
                afterInternalLinkProcessor,
                afterLatexProcessor,
            );
            this.log(
                "afterFormatProcessor",
                afterLatexProcessor,
                afterFormatProcessor,
            );
            this.log(
                "afterFragmentProcessor",
                afterFormatProcessor,
                afterFragmentProcessor,
            );
            this.log(
                "afterDropProcessor",
                afterFragmentProcessor,
                afterDropProcessor,
            );
            this.log(
                "afterGridProcessor",
                afterDropProcessor,
                afterGridProcessor,
            );
            this.log(
                "afterCommentProcessor",
                afterGridProcessor,
                afterCommentProcessor,
            );
            this.log(
                "afterChartProcessor",
                afterCommentProcessor,
                afterChartProcessor,
            );
        }

        return afterChartProcessor;
    }

    postProcess = (element: HTMLElement) => {
        const paragraphs = element.querySelectorAll<HTMLParagraphElement>("p");
        for (let index = 0; index < paragraphs.length; index++) {
            const paragraph = paragraphs.item(index);

            if (paragraph.innerText.startsWith(":::")) {
                paragraph.remove();
            }
        }
    };

    trimEnding(markdown: string, options: Options): string {
        const input = `${markdown}\n`;

        let m = new RegExp(options.separator, "gmi").exec(input);
        if (m !== null) {
            const [match] = m;

            if (input.endsWith(match)) {
                return input.substring(0, input.lastIndexOf(match));
            }
        }

        m = new RegExp(options.verticalSeparator, "gmi").exec(input);
        if (m !== null) {
            const [match] = m;

            if (input.endsWith(match)) {
                return input.substring(0, input.lastIndexOf(match));
            }
        }

        return markdown;
    }

    log(name: string, before: string, after: string) {
        if (before !== after) {
            console.debug(`${name}: ${after}`);
        }
    }
}
