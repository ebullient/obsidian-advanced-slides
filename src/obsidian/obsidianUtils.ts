import { readFileSync } from 'fs-extra';
import { App, FileSystemAdapter, resolveSubpath, TFile } from 'obsidian';
import path from 'path';
import {
    SlidesExtendedSettings as SlidesExtendedSettings,
    ImageCollector,
} from '../@types';
import { DISABLED_IMAGE_COLLECTOR } from '../slidesExtended-constants';
import { MarkdownProcessor } from './markdownProcessor';

let instance: ImageCollector = DISABLED_IMAGE_COLLECTOR;
export function getImageCollector(): ImageCollector {
    return instance;
}

function setImageCollector(newInstance: ImageCollector) {
    instance = newInstance;
}

export class ObsidianUtils implements ImageCollector {
    private app: App;
    private fileSystem: FileSystemAdapter;
    private settings: SlidesExtendedSettings;
    private readonly processor: MarkdownProcessor;
    private readonly vaultDir;
    private readonly pluginDir;
    private readonly distDir;
    private readonly exportDir;
    private readonly cssSearchPath;
    private readonly highlightSearchPath;
    private readonly themeSearchPath;
    private readonly htmlTemplateSearchPath;

    private images = new Set<string>();
    private isCollecting = false;

    private yamlRegex = /^---.*?---\n(.*?)($|---)/s;

    constructor(app: App, settings: SlidesExtendedSettings) {
        this.app = app;
        if (!(this.app.vault.adapter instanceof FileSystemAdapter)) {
            throw new Error('Slides Extended is only supported on the desktop');
        }

        this.fileSystem = this.app.vault.adapter;
        this.settings = settings;
        this.vaultDir = this.fileSystem.getBasePath() + '/';
        this.pluginDir = path.join(
            this.vaultDir,
            this.app.vault.configDir,
            'plugins/slides-extended/',
        );
        this.distDir = path.join(this.pluginDir, 'dist/');
        this.exportDir = path.join(
            this.vaultDir,
            this.settings.exportDirectory,
        );

        this.cssSearchPath = [
            path.join(this.pluginDir), // relative to plugin dir
        ];
        this.highlightSearchPath = [
            path.join(this.pluginDir, 'plugin/highlight'),
        ];
        this.themeSearchPath = [
            path.join(this.pluginDir, 'css'), // plugin layouts
            path.join(this.distDir, 'theme'), // reveal.js themes
        ];
        this.htmlTemplateSearchPath = [path.join(this.pluginDir, 'template')];

        if (this.settings.themeDirectory) {
            const vaultThemeDir = path.join(
                this.vaultDir,
                this.settings.themeDirectory,
            );
            this.cssSearchPath.unshift(vaultThemeDir);
            this.highlightSearchPath.unshift(vaultThemeDir);
            this.themeSearchPath.unshift(vaultThemeDir);
            this.htmlTemplateSearchPath.unshift(vaultThemeDir + '/html');
        }

        setImageCollector(this);
        this.processor = new MarkdownProcessor(this);
    }

    get markdownProcessor(): MarkdownProcessor {
        return this.processor;
    }

    get vaultName(): string {
        return this.fileSystem.getName();
    }
    get vaultDirectory(): string {
        return this.vaultDir;
    }

    get pluginDirectory(): string {
        return this.pluginDir;
    }

    get distDirectory(): string {
        return this.distDir;
    }

    get exportDirectory(): string {
        return this.exportDir;
    }

    getLocalCssSearchPath(): string[] {
        return this.cssSearchPath;
    }

    getThemeSearchPath(): string[] {
        return this.themeSearchPath;
    }
    getHighlightSearchPath(): string[] {
        return this.highlightSearchPath;
    }
    getHtmlTemplateSearchPath(): string[] {
        return this.htmlTemplateSearchPath;
    }

    getSettings(): SlidesExtendedSettings {
        return this.settings;
    }

    private getTFile(filename: string): TFile | null {
        if (filename.startsWith('[[') && filename.endsWith(']]')) {
            filename = filename.substring(2, filename.length - 2).trim();
        }

        const expDir = this.settings.exportDirectory.startsWith('/')
            ? this.settings.exportDirectory.substring(1)
            : this.settings.exportDirectory;

        const allFiles = this.app.vault.getFiles();
        const filesNotInExportDir = allFiles.filter(
            item => !item.path.contains(expDir),
        );
        const allHits = filesNotInExportDir.filter(item =>
            item.path.contains(filename),
        );

        let file: TFile = null;

        // Only one match
        if (allHits.length == 1) {
            file = allHits.first();
        }

        // Workaround for Excalidraw images
        if (!file && filename.toLowerCase().endsWith('.excalidraw')) {
            let hit = filesNotInExportDir.filter(x =>
                x.path.contains(filename + '.svg'),
            );
            if (hit) {
                file = hit.first();
            } else {
                hit = filesNotInExportDir.filter(x =>
                    x.path.contains(filename + '.png'),
                );
                if (hit) {
                    file = hit.first();
                }
            }
        }

        // Find file most similar to search term
        if (!file && allHits.length > 1) {
            let score = 0;
            for (const hit of allHits) {
                const currentScore = this.similarity(filename, hit.path);
                if (currentScore > score) {
                    score = currentScore;
                    file = hit;
                }
            }
        }

        return file;
    }

    getAbsolutePath(relativePath: string): string {
        const markdownFile = this.getTFile(relativePath);
        return this.absolute(markdownFile?.path);
    }

    getRelativePath(path: string): string | null {
        if (path == null) {
            return null;
        }
        const file = this.getTFile(path);
        return file?.path;
    }

    absolute(relativePath: string) {
        if (relativePath) return this.fileSystem.getFullPath(relativePath);
        else {
            return null;
        }
    }

    findFile(path: string) {
        let base = '';
        if (!getImageCollector().shouldCollect()) {
            base = '/';
        }
        const file: TFile = this.getTFile(path);
        if (file) {
            return base + file.path;
        } else {
            return path;
        }
    }

    parseFile(filename: string, header: string) {
        const tfile = this.getTFile(filename);

        if (!tfile) {
            return null;
        }

        const absoluteFilePath = this.absolute(tfile?.path);
        const fileContent = readFileSync(absoluteFilePath, {
            encoding: 'utf-8',
        });

        if (header === null) {
            if (this.yamlRegex.test(fileContent)) {
                return this.yamlRegex.exec(fileContent)[1];
            } else {
                return fileContent;
            }
        } else {
            const cache = this.app.metadataCache.getFileCache(tfile);
            const resolved = resolveSubpath(cache, header);

            if (resolved && resolved.start && resolved.start.line != null) {
                if (resolved.end && resolved.end.line != null) {
                    return this.substring(
                        fileContent,
                        resolved.start.line,
                        resolved.start.col,
                        resolved.end.line,
                        resolved.end.col,
                    );
                } else {
                    return this.substring(
                        fileContent,
                        resolved.start.line,
                        resolved.start.col,
                        -1,
                        -1,
                    );
                }
            } else {
                return '![[' + filename + '#' + header + ']]';
            }
        }
    }

    substring(
        input: string,
        startLine: number,
        startColumn: number,
        endLine: number,
        endColumn: number,
    ): string {
        let result = '';
        const lines = input.split('\n');

        let eline = lines.length;
        if (endLine > -1) {
            eline = endLine;
        }

        for (let index = startLine; index <= eline; index++) {
            const line = lines[index];
            if (line != null) {
                if (index == startLine) {
                    result += line.substring(startColumn) + '\n';
                } else if (index == eline) {
                    let endLine = line;
                    if (endColumn > -1) {
                        endLine = line.substring(0, endColumn);
                    }
                    if (endLine.includes('^')) {
                        endLine = endLine.substring(
                            0,
                            endLine.lastIndexOf('^'),
                        );
                    }
                    result += endLine + '\n';
                } else {
                    result += line + '\n';
                }
            }
        }

        return result;
    }

    similarity(s1: string, s2: string): number {
        let longer = s1;
        let shorter = s2;
        if (s1.length < s2.length) {
            longer = s2;
            shorter = s1;
        }
        const longerLength = longer.length;
        if (longerLength == 0) {
            return 1.0;
        }
        return (
            (longerLength - this.editDistance(longer, shorter)) / longerLength
        );
    }

    editDistance(s1: string, s2: string): number {
        s1 = s1.toLowerCase();
        s2 = s2.toLowerCase();

        const costs = [];
        for (let i = 0; i <= s1.length; i++) {
            let lastValue = i;
            for (let j = 0; j <= s2.length; j++) {
                if (i == 0) costs[j] = j;
                else {
                    if (j > 0) {
                        let newValue = costs[j - 1];
                        if (s1.charAt(i - 1) != s2.charAt(j - 1))
                            newValue =
                                Math.min(
                                    Math.min(newValue, lastValue),
                                    costs[j],
                                ) + 1;
                        costs[j - 1] = lastValue;
                        lastValue = newValue;
                    }
                }
            }
            if (i > 0) costs[s2.length] = lastValue;
        }
        return costs[s2.length];
    }

    resetImageCollection() {
        console.debug(
            'enable image collection for exported slides, start collecting',
        );
        this.images.clear();
        this.isCollecting = true;
    }

    disableImageCollection() {
        console.debug('stop collecting images', this.images);
        this.isCollecting = false;
    }

    // ImageCollector
    addImage(filePath: string): void {
        this.images.add(filePath);
    }

    getAll(): string[] {
        return Array.of(...this.images);
    }

    shouldCollect(): boolean {
        return this.isCollecting;
    }
}
