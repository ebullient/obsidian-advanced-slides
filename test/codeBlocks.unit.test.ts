import { MarkdownProcessor } from 'src/obsidian/markdownProcessor';
import { obsidianUtils as utilsInstance } from './__mocks__/mockObsidianUtils';
import { prepare } from './testUtils';

test('Basic Markdown Syntax > Code Blocks', () => {
	const input = `
\`\`\`dockerfile
FROM ubuntu
\`\`\`
`;

	const { options, markdown } = prepare(input);
	const sut = new MarkdownProcessor(utilsInstance);

	return expect(sut.process(markdown, options)).toMatchSnapshot();
});

test('Basic Markdown Syntax > Code Blocks with multiple $ characters', () => {
	const input = `
\`\`\`dockerfile
USER $USER_NAME:$USER_NAME
\`\`\`
`;

	const { options, markdown } = prepare(input);
	const sut = new MarkdownProcessor(utilsInstance);

	return expect(sut.process(markdown, options)).toMatchSnapshot();
});

test('Code Block Syntax > Headers', () => {
	const input = `
\`\`\`
USER $USER_NAME:$USER_NAME
\`\`\`

The above does not show backticks
`;

	const { options, markdown } = prepare(input);
	const sut = new MarkdownProcessor(utilsInstance);

	return expect(sut.process(markdown, options)).toMatchSnapshot();
});

test('Code Block Syntax > Headers', () => {
	const input = ` \`\`\`
USER $USER_NAME:$USER_NAME
\`\`\`

The above does show backticks
`;

	const { options, markdown } = prepare(input);
	const sut = new MarkdownProcessor(utilsInstance);

	return expect(sut.process(markdown, options)).toMatchSnapshot();
});

test('Code Block Syntax > Headers', () => {
	const input = `\`\`\`
USER $USER_NAME:$USER_NAME
\`\`\`

The above does not show backticks
`;

	const { options, markdown } = prepare(input);
	const sut = new MarkdownProcessor(utilsInstance);

	return expect(sut.process(markdown, options)).toMatchSnapshot();
});

test('Basic Markdown Syntax > Math with Code Blocks', () => {
	const input = `$$\begin{vmatrix}a & b\\
c & d
\end{vmatrix}=ad-bc$$

\`\`\`dockerfile
USER $USER_NAME:$USER_NAME
\`\`\`

You can also do inline math like $s^{-2}_{n}sum_{i=1}^{n}$`;

	const { options, markdown } = prepare(input);
	const sut = new MarkdownProcessor(utilsInstance);

	return expect(sut.process(markdown, options)).toMatchSnapshot();
});

test('Basic Markdown Syntax > Math with Multiple Code Blocks', () => {
	const input = `$$\begin{vmatrix}a & b\\
c & d
\end{vmatrix}=ad-bc$$

\`\`\`dockerfile
USER $USER_NAME:$USER_NAME
\`\`\`

You can also do inline math like $s^{-2}_{n}sum_{i=1}^{n}$

\`\`\`bash
eval "$(/home/$USER_NAME/.rbenv/bin/rbenv init -)"
\`\`\`

That was ruby, now we have javascript:

\`\`\`
console.log("Hello world!")
\`\`\`
`;

	const { options, markdown } = prepare(input);
	const sut = new MarkdownProcessor(utilsInstance);

	return expect(sut.process(markdown, options)).toMatchSnapshot();
});

test('Embedded code has extra characters near dollar signs', () => {
	const input = `
\`\`\`javascript
export class CounterService {
  count$ = new BehaviorSubject(1000);

  double$ = this.count$.pipe(map((count) => count * 2));
  triple$ = this.count$.pipe(map((count) => count * 3));

  combined$ = combineLatest([this.double$, this.triple$]).pipe(
    map(([double, triple]) => double + triple)
  );

  over9000$ = this.combined$.pipe(map((combined) => combined > 9000));

  message$ = this.over9000$.pipe(
    map((over9000) => (over9000 ? "It's over 9000!" : "It's under 9000."))
  );
}
\`\`\``;

	const { options, markdown } = prepare(input);
	const sut = new MarkdownProcessor(utilsInstance);

	return expect(sut.process(markdown, options)).toMatchSnapshot();
});

