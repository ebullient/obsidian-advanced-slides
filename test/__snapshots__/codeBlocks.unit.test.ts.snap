// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Markdown Syntax > Code Blocks 1`] = `
"<!-- .slide: class="drop" -->
<div class="" style="position: absolute; left: 0px; top: 0px; height: 700px; width: 960px; min-height: 700px; display: flex; flex-direction: column; align-items: center; justify-content: center" absolute="true">

\`\`\`dockerfile
FROM ubuntu
\`\`\`
</div>"
`;

exports[`Basic Markdown Syntax > Code Blocks with multiple $ characters 1`] = `
"<!-- .slide: class="drop" -->
<div class="" style="position: absolute; left: 0px; top: 0px; height: 700px; width: 960px; min-height: 700px; display: flex; flex-direction: column; align-items: center; justify-content: center" absolute="true">

\`\`\`dockerfile
USER $USER_NAME:$USER_NAME
\`\`\`
</div>"
`;

exports[`Basic Markdown Syntax > Math with Code Blocks 1`] = `
"<!-- .slide: class="drop" -->
<div class="" style="position: absolute; left: 0px; top: 0px; height: 700px; width: 960px; min-height: 700px; display: flex; flex-direction: column; align-items: center; justify-content: center" absolute="true">

$egin{vmatrix}a & b\\
c & d
end{vmatrix}=ad-bc$

\`\`\`dockerfile
USER $USER_NAME:$USER_NAME
\`\`\`

You can also do inline math like \`$s^{-2}_{n}sum_{i=1}^{n}$\`
</div>"
`;

exports[`Basic Markdown Syntax > Math with Multiple Code Blocks 1`] = `
"<!-- .slide: class="drop" -->
<div class="" style="position: absolute; left: 0px; top: 0px; height: 700px; width: 960px; min-height: 700px; display: flex; flex-direction: column; align-items: center; justify-content: center" absolute="true">

$egin{vmatrix}a & b\\
c & d
end{vmatrix}=ad-bc$

\`\`\`dockerfile
USER $USER_NAME:$USER_NAME
\`\`\`

You can also do inline math like \`$s^{-2}_{n}sum_{i=1}^{n}$\`

\`\`\`bash
eval "$(/home/$USER_NAME/.rbenv/bin/rbenv init -)"
\`\`\`

That was ruby, now we have javascript:

\`\`\`
console.log("Hello world!")
\`\`\`
</div>"
`;

exports[`Code Block Syntax > Headers 1`] = `
"<!-- .slide: class="drop" -->
<div class="" style="position: absolute; left: 0px; top: 0px; height: 700px; width: 960px; min-height: 700px; display: flex; flex-direction: column; align-items: center; justify-content: center" absolute="true">

\`\`\`
USER $USER_NAME:$USER_NAME
\`\`\`

The above does not show backticks
</div>"
`;

exports[`Code Block Syntax > Headers 2`] = `
"<!-- .slide: class="drop" -->
<div class="" style="position: absolute; left: 0px; top: 0px; height: 700px; width: 960px; min-height: 700px; display: flex; flex-direction: column; align-items: center; justify-content: center" absolute="true">

\`\`\`
USER \`$USER_NAME:$\`USER_NAME
\`\`\`

The above does show backticks
</div>"
`;

exports[`Code Block Syntax > Headers 3`] = `
"<!-- .slide: class="drop" -->
<div class="" style="position: absolute; left: 0px; top: 0px; height: 700px; width: 960px; min-height: 700px; display: flex; flex-direction: column; align-items: center; justify-content: center" absolute="true">

\`\`\`
USER $USER_NAME:$USER_NAME
\`\`\`

The above does not show backticks
</div>"
`;

exports[`Embedded code has extra characters near dollar signs 1`] = `
"<!-- .slide: class="drop" -->
<div class="" style="position: absolute; left: 0px; top: 0px; height: 700px; width: 960px; min-height: 700px; display: flex; flex-direction: column; align-items: center; justify-content: center" absolute="true">

\`\`\`javascript
export class CounterService {
  count$ = new BehaviorSubject(1000);

  double$ = this.count$.pipe(map((count) => count * 2));
  triple$ = this.count$.pipe(map((count) => count * 3));

  combined$ = combineLatest([this.double$, this.triple$]).pipe(
    map(([double, triple]) => double + triple)
  );

  over9000$ = this.combined$.pipe(map((combined) => combined > 9000));

  message$ = this.over9000$.pipe(
    map((over9000) => (over9000 ? "It's over 9000!" : "It's under 9000."))
  );
}
\`\`\`
</div>"
`;
